New command logger logcm as a replacement of print calls (updated)

t-woerner opened the pull request:
New command logger logcm as a replacement of print calls

The use of print calls in the installer parts are resulting in the need to
redirect stdout using redirect_stdout in ansible-freeipa. redirect_stdout
is creating a new scope.

The CommandOutput type is a wrapper for a logger instance. It marks log
calls to be printed out on stdout using logger.INFO level. The extra
argument { "cmd_output": True } is set to make sure that a filter could
be used to filter this out for the installer log files.

Usage:

    import logging
    from ipapython.ipa_log_manager import CommandOutput

    logger = logging.getLogger(__name__)
    logcm = CommandOutput(logger)

    def install():
        logcm("Installing ...")

Design note
-----------

The command logger instance is called ``logcm``, because the name has the
same length as the ``print`` function. It may not be the prettiest name,
but it avoids code reformatting and therefore merge conflicts.

At first we considered to add a ``log_cmd`` method to all loggers.
The ``logging.setLoggerClass`` is the public function to set a new logger
class. To make it work, is has to be called before the first logger
objects is instantiated. It's not appropriate to modify the logger
early in ``ipalib``, because ``ipalib`` is also designed to get embedded
into applications. And monkey-patching the ``logging.Logger`` class is
a hack.

The log format for CommandOutput is defined in LOGGING_FORMAT_CMD_OUTPUT
and is applied with the extended Formatter class for the log level INFO
only.

abbra commented:
@t-woerner thanks for the patch. I have one small suggestion: in order to simplify maintenance backports, please split commits that convert individual files to separate commits. This way, for example, a commit that changes `ipaserver/install/server/upgrade.py` wouldn't create a conflict for changes in `ipaserver/install/dsinstance.py` and so on. With smaller commits backports will be automatic.

tiran commented on ipapython/ipa_log_manager.py:
```diff
@@ -197,6 +250,12 @@ def convert_log_level(value):
 
+# Add info_cmd to logger and root logger
+if not hasattr(logging.getLoggerClass(), "info_cmd"):
+    logging.getLoggerClass().info_cmd = log_info_cmd
```
You are suppose to use [logging.setLoggerClass](https://docs.python.org/3/library/logging.html#logging.setLoggerClass) to define your own logger subclass.

t-woerner replied:
This is not needed with the latest version anymore.

tiran commented on ipapython/ipa_log_manager.py:
```diff
@@ -54,6 +54,20 @@
+
+
+# Additional info_log level methods
+def log_info_cmd(obj, msg, *args, **kwargs):
```
Please use self to indicate it's going to be a method:

``def log_info_cmd(self, msg, *args, **kwargs)``

t-woerner replied:
This is used any more with the latest version.

tiran commented on ipapython/ipa_log_manager.py:
```diff
@@ -54,6 +54,20 @@
+# Additional info_log level methods
+def log_info_cmd(obj, msg, *args, **kwargs):
+    kwargs.setdefault("extra", {})["cmd_output"] = True
+    obj._log(logging.INFO, msg, args, **kwargs)
```
Should the level be configurable with a default value of ``INFO``?

t-woerner replied:
This is used any more with the latest version.

tiran commented on ipapython/ipa_log_manager.py:
```diff
@@ -54,6 +54,20 @@
+    obj._log(logging.INFO, msg, args, **kwargs)
+
+
+def log_info_cmd_root(msg, *args, **kwargs):
```
We shouldn't use the root logger at all. If some module uses the root logger, then please fix the module instead.

t-woerner replied:
This is used any more with the latest version.

tiran commented on ipapython/ipa_log_manager.py:
```diff
@@ -141,12 +160,45 @@ def filter(self, record):
+    """
+
+    def filter(self, record):
+        if hasattr(record, "cmd_output") and record.__dict__["cmd_output"]:
```
This can be simplified to ``return getattr(record, "cmd_output", False)``. All extras are available as attributes on the record object.

t-woerner replied:
It can be simplified to `return not getattr(record, "cmd_output", False)`.

tiran commented on ipapython/ipa_log_manager.py:
```diff
@@ -141,12 +160,45 @@ def filter(self, record):
+        Uses LOGGING_FORMAT_INFO_CMD for info_cmd, otherwise normal logging
+        format. For info_cmd record.__dict__["cmd_output"] is set to True.
+        """
+        if record.levelno == logging.INFO and \
```
IMHO we want to support cmd_output for all log levels >= INFO.

``if record.levelno >= logging.INFO and getattr(record, "cmd_output", False):``

t-woerner replied:
Fixed

tiran commented on ipapython/ipa_log_manager.py:
```diff
@@ -141,12 +160,45 @@ def filter(self, record):
+            # Temporarily use LOGGING_FORMAT_INFO_CMD
+            self._fmt = LOGGING_FORMAT_INFO_CMD
+            # Create result
+            result = logging.Formatter.format(self, record)
```
```
try:
    return super(Formatter, self).format(record)
finally:
    self._fmt = orig_fmt
```

t-woerner replied:
Solved in a different way with an special internal Formatter for cmd_output.

tiran commented on ipapython/ipa_log_manager.py:
```diff
@@ -141,12 +160,45 @@ def filter(self, record):
+            # Save original logging format
+            orig_fmt = self._fmt
+            # Temporarily use LOGGING_FORMAT_INFO_CMD
+            self._fmt = LOGGING_FORMAT_INFO_CMD
```
Why is this hack necessary? Formatters are defined on the output handler. We should have a different output handler for stdout and for log files.

Also this is not thread safe. Handlers are shared across threads and multiple threads can log at the same time.

tiran replied:
Update: This does not work with Python 3. The logging module uses a different approach with style formatters and ``formatMessage(self, record)``

t-woerner replied:
Solved by using an internal Formatter for cmd_output.

tiran commented on ipapython/ipa_log_manager.py:
```diff
@@ -197,6 +250,12 @@ def convert_log_level(value):
+if not hasattr(logging.getLoggerClass(), "info_cmd"):
+    logging.getLoggerClass().info_cmd = log_info_cmd
+if not hasattr(logging, "info_cmd"):
+    logging.info_cmd = log_info_cmd_root
```
This should not be necessary.

t-woerner replied:
This is used any more with the latest version.

tiran commented on ipaclient/install/client.py:
```diff
@@ -2055,9 +2055,9 @@ def install_check(options):
-    print("Version {}".format(version.VERSION))
-    print("")
+    logger.info_cmd("This program will set up FreeIPA client.")
+    logger.info_cmd("Version {}".format(version.VERSION))
```
pylint is going to complain about this style. You have to use old-style formatting and replace this kind of calls with ``logger.info_cmd("Version %s", version.VERSION)``.

t-woerner replied:
Fixed

tiran commented on ipalib/cli.py:
```diff
@@ -200,7 +200,7 @@ def print_plain(self, string):
         Print exactly like ``print`` statement would.
         """
-        print(unicode(string))
+        logger.info_cmd(unicode(string))
```
While you are on it, kill ``unicode()`` in ``logger.info_cmd()`` commands

t-woerner replied:
cli.py is not patched anymore

tiran commented on ipalib/messages.py:
```diff
@@ -502,8 +505,8 @@ def iter_messages(variables, base):
     for cls in classes:
-        print('%d\t%s' % (cls.errno, cls.__name__))
-    print('(%d %s)' % (len(classes), label))
+        logger.info_cmd('%d\t%s' % (cls.errno, cls.__name__))
```
format

t-woerner replied:
messages.py is not patched anymore

tiran commented on ipaserver/install/adtrust.py:
```diff
@@ -358,10 +362,11 @@ def install_check(standalone, options, api):
-              "You will not be able to establish trusts with Active "
-              "Directory unless\nthe realm name of the IPA server matches its "
-              "domain name.\n\n")
+        logger.info_cmd(
```
For this kind of stuff it should be possible to increase the log level to WARNING.

t-woerner replied:
I think this should be an extra commit.

t-woerner committed:
ipaclient/install/client: Use logcm as a replacement of print calls

The CommandOutput type is a wrapper for a logger instance. It marks log
calls to be printed out on stdout using logger.INFO level. The extra
argument { "cmd_output": True } is set to make sure that a filter could
be used to filter this out for the installer log files.

A new import of CommandOutput has been added and a call to create logcm.

Additonally the import of print_function has been removed because print
is not used anymore.

Fixes: https://pagure.io/freeipa/issue/7914
Signed-off-by: Thomas Woerner <twoerner@redhat.com>

tiran commented:
> @t-woerner thanks for the patch. I have one small suggestion: in order to simplify maintenance backports, please split commits that convert individual files to separate commits. This way, for example, a commit that changes `ipaserver/install/server/upgrade.py` wouldn't create a conflict for changes in `ipaserver/install/dsinstance.py` and so on. With smaller commits backports will be automatic.

I don't understand how this approach is going to reduce the amount of merge conflicts in backports. How do multiple commits reduce the amount of conflicts when none of the commits are cherry-picked into the backport branch?

tiran commented:
May I propose a slightly different approach that avoids monkey patching of classes and reduces the amount of code reformatting?

```
import logging

class CommandOutput:
    """Command output -- log and print to user

    The CommandOutput type is a wrapper for a logger instance. It
    marks log calls to be printed out on stdout.

    Usage:

        import logging
        from ipapython.ipa_log_manager import CommandOutput

        logger = logging.getLogger(__name__)
        logcm = CommandOutput(logger)

        def install():
            logcm("Installing ...")

    Design note
    -----------

    The command logger instance is called ``logcm``, because the name has the
    same length as the ``print`` function. It may not be the prettiest name,
    but it avoids code reformatting and therefore merge conflicts.

    At first we considered to add a ``log_cmd`` method to all loggers.
    The ``logging.setLoggerClass`` is the public function to set a new logger
    class. To make it work, is has to be called before the first logger
    objects is instantiated. It's not appropriate to modify the logger
    early in ``ipalib``, because ``ipalib`` is also designed to get embedded
    into applications. And monkey-patching the ``logging.Logger`` class is
    a hack.
    """
    __slots__ = ('_logger',)

    def __init__(self, logger):
        self._logger = logger

    def __call__(self, msg, *args, **kwargs):
        if 'extra' not in kwargs:
            kwargs["extra"] = {"cmd_output": True}
        else:
            kwargs["extra"]["cmd_output"] = True
        level = kwargs.pop("level", logging.INFO)
        self._logger.log(level, msg, *args, **kwargs)
```

tiran commented on ipalib/cli.py:
```diff
@@ -200,7 +200,7 @@ def print_plain(self, string):
         Print exactly like ``print`` statement would.
         """
-        print(unicode(string))
+        logger.info_cmd(unicode(string))
```
``ipalib.run`` implements the ``ipa`` command line tool. You have to keep all ``print()`` calls.

t-woerner replied:
Fixed

tiran commented:
Please only modify the installer code and unpatch the following files.

```
ipalib/cli.py
ipalib/messages.py
ipaserver/dnssec/ldapkeydb.py
ipaserver/dnssec/localhsm.py
```

``ipalib.cli`` implements the ``ipa`` command. With the current state, ``ipa`` does not print anything. The other modules use ``print()`` in ``__main__`` blocks, which are only used for development and debugging.

abbra commented:
> I don't understand how this approach is going to reduce the amount of merge conflicts in backports. How do multiple commits reduce the amount of conflicts when none of the commits are cherry-picked into the backport branch?

Suppose you have to backport a fix that was committed after these changes landed, in a proximity of the changes. When the changes split into smaller commits, a backport would result in porting a smaller base commit that adds logging support and the small patch from the affected area. If this would have been a single commit that touches multiple files, you would have to create a manual backport that will involve quite a lot of effort to make sure all pieces are not conflicting with something else.

t-woerner committed:
ipaserver/install/adtrust: Use logcm as a replacement of print calls

The CommandOutput type is a wrapper for a logger instance. It marks log
calls to be printed out on stdout using logger.INFO level. The extra
argument { "cmd_output": True } is set to make sure that a filter could
be used to filter this out for the installer log files.

A new import of CommandOutput has been added and a call to create logcm.

Additonally the import of print_function has been removed because print
is not used anymore.

Fixes: https://pagure.io/freeipa/issue/7914
Signed-off-by: Thomas Woerner <twoerner@redhat.com>

t-woerner committed:
ipaserver/install/adtrustinstance: Use logcm as a replacement of print calls

The CommandOutput type is a wrapper for a logger instance. It marks log
calls to be printed out on stdout using logger.INFO level. The extra
argument { "cmd_output": True } is set to make sure that a filter could
be used to filter this out for the installer log files.

A new import of CommandOutput has been added and a call to create logcm.

Additonally the import of print_function has been removed because print
is not used anymore.

Fixes: https://pagure.io/freeipa/issue/7914
Signed-off-by: Thomas Woerner <twoerner@redhat.com>

t-woerner committed:
ipaserver/install/bindinstance: Use logcm as a replacement of print calls

The CommandOutput type is a wrapper for a logger instance. It marks log
calls to be printed out on stdout using logger.INFO level. The extra
argument { "cmd_output": True } is set to make sure that a filter could
be used to filter this out for the installer log files.

A new import of CommandOutput has been added and a call to create logcm.

Additonally the import of print_function has been removed because print
is not used anymore.

Fixes: https://pagure.io/freeipa/issue/7914
Signed-off-by: Thomas Woerner <twoerner@redhat.com>

t-woerner committed:
ipaserver/install/ca: Use logcm as a replacement of print calls

The CommandOutput type is a wrapper for a logger instance. It marks log
calls to be printed out on stdout using logger.INFO level. The extra
argument { "cmd_output": True } is set to make sure that a filter could
be used to filter this out for the installer log files.

A new import of CommandOutput has been added and a call to create logcm.

Additonally the import of print_function has been removed because print
is not used anymore and formatting using .format has been replaced by old
style formatting.

Fixes: https://pagure.io/freeipa/issue/7914
Signed-off-by: Thomas Woerner <twoerner@redhat.com>

t-woerner committed:
ipaserver/install/custodiainstance: Use logcm as a replacement of print calls

The CommandOutput type is a wrapper for a logger instance. It marks log
calls to be printed out on stdout using logger.INFO level. The extra
argument { "cmd_output": True } is set to make sure that a filter could
be used to filter this out for the installer log files.

A new import of CommandOutput has been added and a call to create logcm.

Additonally the import of print_function has been removed because print
is not used anymore.

Fixes: https://pagure.io/freeipa/issue/7914
Signed-off-by: Thomas Woerner <twoerner@redhat.com>

t-woerner committed:
ipaserver/install/dns: Use logcm as a replacement of print calls

The CommandOutput type is a wrapper for a logger instance. It marks log
calls to be printed out on stdout using logger.INFO level. The extra
argument { "cmd_output": True } is set to make sure that a filter could
be used to filter this out for the installer log files.

A new import of CommandOutput has been added and a call to create logcm.

Additonally the import of print_function has been removed because print
is not used anymore.

Fixes: https://pagure.io/freeipa/issue/7914
Signed-off-by: Thomas Woerner <twoerner@redhat.com>

t-woerner committed:
ipaserver/install/dnskeysyncinstance: Use logcm as a replacement of print calls

The CommandOutput type is a wrapper for a logger instance. It marks log
calls to be printed out on stdout using logger.INFO level. The extra
argument { "cmd_output": True } is set to make sure that a filter could
be used to filter this out for the installer log files.

A new import of CommandOutput has been added and a call to create logcm.

Additonally the import of print_function has been removed because print
is not used anymore.

Fixes: https://pagure.io/freeipa/issue/7914
Signed-off-by: Thomas Woerner <twoerner@redhat.com>

t-woerner committed:
ipaserver/install/dsinstance: Use logcm as a replacement of print calls

The CommandOutput type is a wrapper for a logger instance. It marks log
calls to be printed out on stdout using logger.INFO level. The extra
argument { "cmd_output": True } is set to make sure that a filter could
be used to filter this out for the installer log files.

A new import of CommandOutput has been added and a call to create logcm.

Additonally the import of print_function has been removed because print
is not used anymore.

Fixes: https://pagure.io/freeipa/issue/7914
Signed-off-by: Thomas Woerner <twoerner@redhat.com>

t-woerner committed:
ipaserver/install/installutils: Use logcm as a replacement of print calls

The CommandOutput type is a wrapper for a logger instance. It marks log
calls to be printed out on stdout using logger.INFO level. The extra
argument { "cmd_output": True } is set to make sure that a filter could
be used to filter this out for the installer log files.

A new import of CommandOutput has been added and a call to create logcm.

Fixes: https://pagure.io/freeipa/issue/7914
Signed-off-by: Thomas Woerner <twoerner@redhat.com>

t-woerner committed:
ipaserver/install/ipa_crlgen_manage: Use logcm as a replacement of print calls

The CommandOutput type is a wrapper for a logger instance. It marks log
calls to be printed out on stdout using logger.INFO level. The extra
argument { "cmd_output": True } is set to make sure that a filter could
be used to filter this out for the installer log files.

A new import of CommandOutput has been added and a call to create logcm.

Additonally the import of print_function has been removed because print
is not used anymore.

Fixes: https://pagure.io/freeipa/issue/7914
Signed-off-by: Thomas Woerner <twoerner@redhat.com>

t-woerner committed:
ipaserver/install/replication: Use logcm as a replacement of print calls

The CommandOutput type is a wrapper for a logger instance. It marks log
calls to be printed out on stdout using logger.INFO level. The extra
argument { "cmd_output": True } is set to make sure that a filter could
be used to filter this out for the installer log files.

A new import of CommandOutput has been added and a call to create logcm.

Additonally the import of print_function has been removed because print
is not used anymore.

Fixes: https://pagure.io/freeipa/issue/7914
Signed-off-by: Thomas Woerner <twoerner@redhat.com>

t-woerner committed:
ipaserver/install/server/__init__: Use logcm as a replacement of print calls

The CommandOutput type is a wrapper for a logger instance. It marks log
calls to be printed out on stdout using logger.INFO level. The extra
argument { "cmd_output": True } is set to make sure that a filter could
be used to filter this out for the installer log files.

A new import of CommandOutput has been added and a call to create logcm.

Additonally the import of print_function has been removed because print
is not used anymore.

Fixes: https://pagure.io/freeipa/issue/7914
Signed-off-by: Thomas Woerner <twoerner@redhat.com>

t-woerner committed:
ipaserver/install/server/install: Use logcm as a replacement of print calls

The CommandOutput type is a wrapper for a logger instance. It marks log
calls to be printed out on stdout using logger.INFO level. The extra
argument { "cmd_output": True } is set to make sure that a filter could
be used to filter this out for the installer log files.

A new import of CommandOutput has been added and a call to create logcm.

Additonally the import of print_function has been removed because print
is not used anymore.

Fixes: https://pagure.io/freeipa/issue/7914
Signed-off-by: Thomas Woerner <twoerner@redhat.com>

t-woerner committed:
ipaserver/install/server/replicainstall: Use logcm as a replacement of print calls

The CommandOutput type is a wrapper for a logger instance. It marks log
calls to be printed out on stdout using logger.INFO level. The extra
argument { "cmd_output": True } is set to make sure that a filter could
be used to filter this out for the installer log files.

A new import of CommandOutput has been added and a call to create logcm.

Fixes: https://pagure.io/freeipa/issue/7914
Signed-off-by: Thomas Woerner <twoerner@redhat.com>

t-woerner committed:
ipaserver/install/server/upgrade: Use logcm as a replacement of print calls

The CommandOutput type is a wrapper for a logger instance. It marks log
calls to be printed out on stdout using logger.INFO level. The extra
argument { "cmd_output": True } is set to make sure that a filter could
be used to filter this out for the installer log files.

A new import of CommandOutput has been added and a call to create logcm.

Additonally the import of print_function has been removed because print
is not used anymore.

Fixes: https://pagure.io/freeipa/issue/7914
Signed-off-by: Thomas Woerner <twoerner@redhat.com>

tiran reviewed (changes requested):
There should be no no calls to ``.format()`` or `` % `` in ``logcm()``. I marked some of the calls, but there are many more.

t-woerner commented on ipaclient/install/client.py:
```diff
@@ -2055,9 +2056,9 @@ def install_check(options):
-    print("Version {}".format(version.VERSION))
-    print("")
+    logcm("This program will set up FreeIPA client.")
+    logcm("Version %s" % version.VERSION)
```
Done

t-woerner replied:
Done

t-woerner commented on ipaserver/install/custodiainstance.py:
```diff
@@ -227,10 +229,9 @@ def _wait_keys(self):
-                            self.ldap_uri)
+                    logcm(
+                        "  Waiting for keys to appear on host: %s, please "
+                        "wait until this has completed." % self.ldap_uri
```
Done

t-woerner replied:
Done

t-woerner commented on ipaserver/install/dns.py:
```diff
@@ -216,7 +222,7 @@ def install_check(standalone, api, replica, options, hostname):
         # we can reinstall current server if it is dnssec master
         if dnssec_masters and api.env.host not in dnssec_masters:
-            print("DNSSEC key master(s):", u','.join(dnssec_masters))
+            logcm("DNSSEC key master(s): %s" % u','.join(dnssec_masters))
```
Done

t-woerner replied:
Done

t-woerner commented on ipaserver/install/dsinstance.py:
```diff
@@ -1052,7 +1054,7 @@ def change_admin_password(self, password):
                 logger.debug("ldappasswd done")
             except ipautil.CalledProcessError as e:
-                print("Unable to set admin password", e)
+                logcm("Unable to set admin password %s" % e)
```
Done

t-woerner commented on ipaserver/install/installutils.py:
```diff
@@ -175,7 +177,7 @@ def verify_fqdn(host_name, no_host_dns=False, local_hostname=True):
 
     if no_host_dns:
-        print("Warning: skipping DNS resolution of host", host_name)
+        logcm("Warning: skipping DNS resolution of host %s" % host_name)
```
Done

t-woerner commented on ipaserver/install/installutils.py:
```diff
@@ -278,7 +280,7 @@ def read_ip_addresses():
             ip_parsed = ipautil.CheckedIPAddress(ip)
         except Exception as e:
-            print("Error: Invalid IP Address %s: %s" % (ip, e))
+            logcm("Error: Invalid IP Address %s: %s" % (ip, e))
```
Done

abbra commented on ipaclient/install/client.py:
```diff
@@ -2055,9 +2056,9 @@ def install_check(options):
-    print("Version {}".format(version.VERSION))
-    print("")
+    logcm("This program will set up FreeIPA client.")
+    logcm("Version %s", version.VERSION)
```
Wrong conversion here, you need `%`, not `,`

tiran replied:
no, that's the correct form.

abbra commented on ipaclient/install/client.py:
```diff
@@ -2084,10 +2085,10 @@ def install_check(options):
-            print("in favor of chronyd")
-            print("")
+            logcm("WARNING: conflicting time&date synchronization service '%s'"
+                  " will be disabled", e.conflicting_service)
```
Same here

tiran replied:
No, Thomas is using the recommended form here. See https://docs.python.org/3/library/logging.html#logging.debug

abbra commented on ipaserver/install/custodiainstance.py:
```diff
@@ -227,10 +229,9 @@ def _wait_keys(self):
-                            self.ldap_uri)
+                    logcm(
+                        "  Waiting for keys to appear on host: %s, please "
+                        "wait until this has completed.", self.ldap_uri
```
Same here

abbra commented on ipaserver/install/dsinstance.py:
```diff
@@ -1052,7 +1054,7 @@ def change_admin_password(self, password):
                 logger.debug("ldappasswd done")
             except ipautil.CalledProcessError as e:
-                print("Unable to set admin password", e)
+                logcm("Unable to set admin password %s", e)
```
Same here

abbra commented on ipaserver/install/installutils.py:
```diff
@@ -254,7 +256,7 @@ def record_in_hosts(ip, host_name=None, conf_file=paths.HOSTS):
             return (hosts_ip, names)
         except IndexError:
-            print("Warning: Erroneous line '%s' in %s" % (line, conf_file))
+            logcm("Warning: Erroneous line '%s' in %s", line, conf_file)
```
same here

abbra commented on ipaserver/install/installutils.py:
```diff
@@ -278,7 +280,7 @@ def read_ip_addresses():
             ip_parsed = ipautil.CheckedIPAddress(ip)
         except Exception as e:
-            print("Error: Invalid IP Address %s: %s" % (ip, e))
+            logcm("Error: Invalid IP Address %s: %s", ip, e)
```
Same here

abbra commented on ipaserver/install/installutils.py:
```diff
@@ -288,12 +290,12 @@ def read_ip_addresses():
     addrs = []
     if ipautil.user_input("Do you want to configure DNS forwarders?", True):
-        print("Following DNS servers are configured in /etc/resolv.conf: %s" %
+        logcm("Following DNS servers are configured in /etc/resolv.conf: %s",
```
same here

abbra commented on ipaserver/install/installutils.py:
```diff
@@ -303,15 +305,15 @@ def read_dns_forwarders():
-                print("Error: Invalid IP Address %s: %s" % (ip, e))
-                print("DNS forwarder %s not added." % ip)
+                logcm("Error: Invalid IP Address %s: %s", ip, e)
+                logcm("DNS forwarder %s not added.", ip)
```
And these two lines

abbra commented on ipaserver/install/installutils.py:
```diff
@@ -303,15 +305,15 @@ def read_dns_forwarders():
                 continue
 
-            print("DNS forwarder %s added. You may add another." % ip)
+            logcm("DNS forwarder %s added. You may add another.", ip)
```
Same here

abbra commented on ipaserver/install/installutils.py:
```diff
@@ -393,7 +395,7 @@ def update_file(filename, orig, subst):
         return 0
     else:
-        print("File %s doesn't exist." % filename)
+        logcm("File %s doesn't exist.", filename)
```
Same here

abbra commented on ipaserver/install/installutils.py:
```diff
@@ -544,8 +546,8 @@ def update_hosts_file(ip_addresses, host_name, fstore):
             continue
-        print("Adding [{address!s} {name}] to your /etc/hosts file".format(
-            address=ip_address, name=host_name))
+        logcm("Adding [%s %s] to your /etc/hosts file", ip_address,
```
Same here

abbra commented:
@t-woerner I found a bunch of same type conversion errors that need to be fixed.

tiran commented:
> @t-woerner I found a bunch of same type conversion errors that need to be fixed.

The code is correct and uses the preferred format for logging. The logger does not perform string formatting by itself. The format message and arguments are passed to handlers and then formatted inside the handler's log formatter.

abbra commented:
@t-woerner @tiran my fault (too late after work, you know..) So please ignore it. ;) LGTM

tiran commented on ipapython/ipa_log_manager.py:
```diff
@@ -141,12 +190,37 @@ def filter(self, record):
+        Uses LOGGING_FORMAT_CMD_OUTPUT for info_cmd, otherwise normal logging
+        format. For info_cmd record.__dict__["cmd_output"] is set to True.
+        """
+        if record.levelno >= logging.INFO and \
```
This logic should not be in the formatter but rather in the handler's formatter configuration. With this approach the debug logs will miss context and time information for ``logcm`` events. Please use a different formatter for log files and stdout logging here:

https://github.com/freeipa/freeipa/blob/b79ea6a6e92d224859c2d4533b99b3c15d3711ca/ipapython/ipa_log_manager.py#L166-L179

https://github.com/freeipa/freeipa/blob/b79ea6a6e92d224859c2d4533b99b3c15d3711ca/ipalib/plugable.py#L450-L522

tiran replied:
This is the wrong approach, see [flow diagram](https://docs.python.org/3/howto/logging.html#logging-flow). The logic should be in the handler configuration. You have to use multiple handlers:

1. One stdout handler with a filter for ``cmd_output`` records, a filter for ``level < WARNING`` records, and a ``cmd_output_format`` formatter.
2. One stderr handler with a filter for ``cmd_output`` records, a filter for ``level >= WARNING`` records, and a ``cmd_output_format`` formatter.
3. One stdout handler with a filter for non-``cmd_output`` records and ``console_format`` formatter.

Handler 3 could be split into two separate handlers. I think it would be a good idea to not print a full traceback in ``cmd_output_formatter`` (see ``formatException``).

tiran commented on ipapython/ipa_log_manager.py:
```diff
@@ -164,6 +238,7 @@ def standard_logging_setup(filename=None, verbose=False, debug=False,
         finally:
             os.umask(umask)
         file_handler.setLevel(logging.DEBUG)
+        file_handler.addFilter(InfoCommandFilter())
```
Why do you add the filter to the file logger instead of the console logger? Isn't the filter suppose to filter which events end up in stdout?

t-woerner replied:
Adding the filter to the installation log fil e handler makes sure that the print output is not in the log file (like it has been up to now) and is by default in the normal installer (log) output. This makes sure that an existing application using FreeIPA parts does not need to enable this explicitly.

tiran replied:
I think we have some kind of misunderstanding. Having the output in the log files is one of the benefits for me. It makes it easier to correlate debug logs with installer console output. 

I proposed the ``cmd_output`` flag to mark log entries that **also** land on stdout (console logger).

t-woerner replied:
This is behavior change that needs to be verified also in all the tests.

tiran commented:
``ipapython.admintool`` also performs logger configuration. I think you need to address ``setup_logging``, too.

t-woerner committed:
New command logger logcm as a replacement of print calls

The use of print calls in the installer parts are resulting in the need to
redirect stdout using redirect_stdout in ansible-freeipa. redirect_stdout
is creating a new scope.

The CommandOutput type is a wrapper for a logger instance. It marks log
calls to be printed out on stdout using logger.INFO level. The extra
argument { "cmd_output": True } is set to make sure that a filter could
be used to filter this out for the installer log files.

Usage:

    import logging
    from ipapython.ipa_log_manager import CommandOutput

    logger = logging.getLogger(__name__)
    logcm = CommandOutput(logger)

    def install():
        logcm("Installing ...")

Design note
-----------

The command logger instance is called ``logcm``, because the name has the
same length as the ``print`` function. It may not be the prettiest name,
but it avoids code reformatting and therefore merge conflicts.

At first we considered to add a ``log_cmd`` method to all loggers.
The ``logging.setLoggerClass`` is the public function to set a new logger
class. To make it work, is has to be called before the first logger
objects is instantiated. It's not appropriate to modify the logger
early in ``ipalib``, because ``ipalib`` is also designed to get embedded
into applications. And monkey-patching the ``logging.Logger`` class is
a hack.

The log format for CommandOutput is defined in LOGGING_FORMAT_CMD_OUTPUT
and is applied with the extended Formatter class for the log level INFO
only.

Fixes: https://pagure.io/freeipa/issue/7914
Signed-off-by: Thomas Woerner <twoerner@redhat.com>

t-woerner committed:
ipaserver/install/cainstance: Use logcm as a replacement of print calls

The CommandOutput type is a wrapper for a logger instance. It marks log
calls to be printed out on stdout using logger.INFO level. The extra
argument { "cmd_output": True } is set to make sure that a filter could
be used to filter this out for the installer log files.

A new import of CommandOutput has been added and a call to create logcm.

Additonally the import of print_function has been removed because print
is not used anymore.

Fixes: https://pagure.io/freeipa/issue/7914
Signed-off-by: Thomas Woerner <twoerner@redhat.com>

t-woerner commented:
> `ipapython.admintool` also performs logger configuration. I think you need to address `setup_logging`, too.

ipapython.admintool._setup_logging, which is used in ipapython.admintool.setup_logging is using standard_logging_setup. Therefore it should not be needed to do further changes here.

tiran reviewed (changes requested):
There are still a bunch of print() calls in the installer code.

``logcm("WARNING: ...")`` and ``print(file=sys.stderr)`` should be replaced with ``logcm(..., level=logging.WARNING)`` and a proper formatter/handler combination. The formatter could prepend ``WARN``.

duzda commented:
Hi @t-woerner, are you still planning to work on this or can this safely be closed?

duzda commented:
Hi, closing due to inactivity, if this ever becomes relevant again, please reopen
